src/main/java/org/springframework/data/elasticsearch/core/SearchRequestFactory.java

```diff
@@ -0,0 +1,326 @@
+
+		if (!query.getFields().isEmpty()) {
+			searchRequestBuilder.setFetchSource(query.getFields().toArray(new String[0]), null);
+		}
+
+		if (query.getIndicesOptions() != null) {
+			searchRequestBuilder.setIndicesOptions(query.getIndicesOptions());
+		}
+
+		if (query.getSort() != null) {
+			prepareSort(query, searchRequestBuilder, entity);
+		}
+
+		if (query.getMinScore() > 0) {
+			searchRequestBuilder.setMinScore(query.getMinScore());
+		}
+
+		if (query.getPreference() != null) {
+			searchRequestBuilder.setPreference(query.getPreference());
+		}
+
+		return searchRequestBuilder;
+	}
+
+	private void prepareSort(Query query, SearchSourceBuilder sourceBuilder,
+			@Nullable ElasticsearchPersistentEntity<?> entity) {
+		for (Sort.Order order : query.getSort()) {
+			sourceBuilder.sort(getSortBuilder(order, entity));
+		}
+	}
+
+	private void prepareSort(Query query, SearchRequestBuilder searchRequestBuilder,
+			@Nullable ElasticsearchPersistentEntity<?> entity) {
+		for (Sort.Order order : query.getSort()) {
+			searchRequestBuilder.addSort(getSortBuilder(order, entity));
+		}
+	}
+
+	private SortBuilder getSortBuilder(Sort.Order order, @Nullable ElasticsearchPersistentEntity<?> entity) {
+		SortOrder sortOrder = order.getDirection().isDescending() ? SortOrder.DESC : SortOrder.ASC;
+
+		if (ScoreSortBuilder.NAME.equals(order.getProperty())) {
+			return SortBuilders //
+					.scoreSort() //
+					.order(sortOrder);
+		} else {
+			ElasticsearchPersistentProperty property = (entity != null) //
+					? entity.getPersistentProperty(order.getProperty()) //
+					: null;
+			String fieldName = property != null ? property.getFieldName() : order.getProperty();
+
+			FieldSortBuilder sort = SortBuilders //
+					.fieldSort(fieldName) //
+					.order(sortOrder);
+
+			if (order.getNullHandling() == Sort.NullHandling.NULLS_FIRST) {
+				sort.missing("_first");
+			} else if (order.getNullHandling() == Sort.NullHandling.NULLS_LAST) {
+				sort.missing("_last");
+			}
+
+			return sort;
+		}
+	}
+
+	private Pair<QueryBuilder, QueryBuilder> getQueryAndFilter(Query query) {
+		QueryBuilder elasticsearchQuery;
+		QueryBuilder elasticsearchFilter;
+
+		if (query instanceof NativeSearchQuery) {
+			NativeSearchQuery searchQuery = (NativeSearchQuery) query;
+			elasticsearchQuery = searchQuery.getQuery();
+			elasticsearchFilter = searchQuery.getFilter();
+		} else if (query instanceof CriteriaQuery) {
+			CriteriaQuery criteriaQuery = (CriteriaQuery) query;
+			elasticsearchQuery = new CriteriaQueryProcessor().createQueryFromCriteria(criteriaQuery.getCriteria());
+			elasticsearchFilter = new CriteriaFilterProcessor().createFilterFromCriteria(criteriaQuery.getCriteria());
+		} else {
+			throw new IllegalArgumentException("unhandled Query implementation " + query.getClass().getName());
+		}
+
+		return new Pair<>(elasticsearchQuery, elasticsearchFilter);
+	}
+
+	@Nullable
+	private String[] retrieveIndexNameFromPersistentEntity(@Nullable Class clazz) {
```

Comment:
The class should be non-nullable and we should use `getRequiredPersistentEntity(…)` to avoid NPEs.

This method is called when we do not have index coord specified, so having a persistent entity is a requirement. Same for `retrieveTypeFromPersistentEntity`.